ENH: Enabled the use of numpy.vectorize as a decorator

MatteoRaso opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->
This closes #9477 and #23021, as part of a small project suggested by @seberg.

MatteoRaso committed:
BUG: Added __name__ atribute to vectorize class (#23021)

MatteoRaso committed:
ENH: Enabled use of numpy.vectorize as decorator (#9477)

Most of this code comes from PR-9593, but with the default value for pyfunc
being numpy._NoValue instead of None, no override of __new__, and no effort
to make subclassing possible.

Co-Authored-By: Michael Lamparski <diagonaldevice@gmail.com>

MatteoRaso committed:
Fixed lint error

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -2119,8 +2119,8 @@ def _create_arrays(broadcast_shape, dim_sizes, list_of_core_dims, dtypes,
     """
-    vectorize(pyfunc, otypes=None, doc=None, excluded=None, cache=False,
-              signature=None)
+    vectorize(pyfunc=np._NoValue, otypes=None, doc=None, excluded=None,
```
Maybe time to replace this with something like "Create a vectorized version of a function, able to process arrays" -- after all, the signature is always shown by python help and the docs.

MatteoRaso replied:
Agreed. Strictly speaking, I don't even think np.vectorize counts as a generalized function class.

MatteoRaso replied:
> Is there a reason not to use `None`?

Using would make ```@vectorize(None)``` a legal call. It's not the worst thing, but it's strange and I'd prefer avoiding it. There was a bit of controversy over this in the old PR, btw.

mhvk replied:
Looking at this again, I think `None` is actually fine here...

MatteoRaso replied:
>Looking at this again, I think None is actually fine here...

I just don't want to risk some edge case where somebody accidentally passes None.

seberg replied:
Eric preferred to not use None in the other PR.  Personally, I still prefer `None` slightly, `NoValue` seems a bit too smart for no reason I can think of where users might pass `None` accidentally.

MatteoRaso committed:
Changed documentation for numpy.vectorize

Previous definition implies that vectorize returns generalized functions,
which is not true.

seberg reviewed (commented):
Could you have a look through why the old PR got closed?  I wonder if there were any API discussions there that are still relevant.

It seems likely that there really should be a `vectorized()` version which is meant to be used as a decorator only, but that should probably also do more changes while at it :/.

If we make this a decorator, I would say the decorator example should come first to encourage its use.  It does need a `.. versionchanged` to indicate the added capability.

MatteoRaso commented:
>Could you have a look through why the old PR got closed? I wonder if there were any API discussions there that are still relevant.

The reason the old PR didn't go through was because of problems dealing with ```__new__``` and trying to make vectorize work well with subclassing, which I intentionally didn't bother with.

MatteoRaso committed:
Added test for vectorize decorator project

This test makes sure a function's docstring is properly preserved
after calling the vectorize decorator.

MatteoRaso commented on numpy/lib/function_base.py:
```diff
@@ -2300,7 +2320,15 @@ def __init__(self, pyfunc, otypes=None, doc=None, excluded=None,
+        if self._doc is None:
+            self.__doc__ = pyfunc.__doc__
+        else:
+            self.__doc__ = self._doc
```
Using ```functools.wraps``` isn't actually necessary here, since we're making a decorator for a class instead of a function. That being said, your comment made me realize that I didn't have a test that checked to make sure the docstring is properly preserved after using @vectorize.

seberg replied:
We have a callable object, so that is basically equivalent to a function.  An argument may be that `__wrapped__` is not strictly true and e.g. inheriting the signature annotations might actually be wrong.  So in that sense it may be better to maybe use it, but pick out the interesting attributes for now, which the `functools.wraps` doc list (or `updated_wrapper`).

When used as a decorator, `__module__` is a very interesting attribute because it is a step towards allowing pickling, and if we use name, qualname is also.

MatteoRaso commented:
Any follow up to this?

mattip commented on numpy/lib/function_base.py:
```diff
@@ -2267,18 +2269,36 @@ class vectorize:
-    def __init__(self, pyfunc, otypes=None, doc=None, excluded=None,
-                 cache=False, signature=None):
+    def __init__(self, pyfunc=np._NoValue, otypes=None, doc=None,
+                 excluded=None, cache=False, signature=None):
```
What happens if the decorator is used with positional args
```
@vectorize(['float64'], "return float absolute value")
def myabs(a):
    return abs(a)
```

vs.
```
myabs = np.vectorize(abs, ['float64'], "return float absolute value")
```

I don't think we can change the signature to require kwargs since that would break existing code
```
def __init__(self, *, pyfunc=np._NoValue, ...
```

MatteoRaso replied:
>What happens if the decorator is used with positional args

Doesn't work, it assigns the list to pyfunc. Not sure how this can be fixed. Type annotations might work, but I feel like there's a better solution.

mhvk replied:
One could check whether the first argument is `callable`? But it would then mean shifting the rest. Perhaps it is OK to insist on keyword arguments if used as a decorator?

MatteoRaso replied:
> But it would then mean shifting the rest

This is what I went for. It solves the issue, but it requires a hideous sequence of nested if-else statements, although it's oddly pep8-compliant.

seberg replied:
My gut feeling would be to just bail and raise an informative error when `pfunc` is not callable.  Allowing `None` in that sense might even be useful, but I don't care, the error could say: if you are using a decorator, you must use kwargs.

That said, if you all like it, its fine.  Just such logic tends to be hairy...

There may be a point on introducing e.g. `vectorized` as a new name or so and trying to move users over, but while I would like it, I feel it should probably come with deeper improvements of `vectorize`.

MatteoRaso replied:
>My gut feeling would be to just bail and raise an informative error when pfunc is not callable. Allowing None in that sense might even be useful, but I don't care, the error could say: if you are using a decorator, you must use kwargs.

I'd do that if there was a consensus about it, but I'd prefer not to. Generally, I don't like to shoehorn the user into doing things like having to specify the keywords in arguments. It'd be inconvenient for the user and likely cause bugs because of users misspelling keywords.

mhvk replied:
I'm with @seberg here -- and keyword arguments would definitely seem better when used inside code, as a decorator, where really one has to worry about readability first.

MatteoRaso replied:
Ok, I changed it.

MatteoRaso commented:
>Perhaps it is OK to insist on keyword arguments if used as a decorator?

If we must, but I'd prefer to figure out a solution rather than push the onus onto
the end user.

------- Original Message -------
On Tuesday, February 7th, 2023 at 9:01 PM, Marten van Kerkwijk ***@***.***> wrote:

> @mhvk commented on this pull request.
>
> ---------------------------------------------------------------
>
> In [numpy/lib/function_base.py](https://github.com/numpy/numpy/pull/23061#discussion_r1099556103):
>
>>      """
> -    def __init__(self, pyfunc, otypes=None, doc=None, excluded=None,
> -                 cache=False, signature=None):
> +    def __init__(self, pyfunc=np._NoValue, otypes=None, doc=None,
> +                 excluded=None, cache=False, signature=None):
>
> One could check whether the first argument is callable? But it would then mean shifting the rest. Perhaps it is OK to insist on keyword arguments if used as a decorator?
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/numpy/numpy/pull/23061#discussion_r1099556103), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AIDGW6XA7LLWM6NGYFY4U6LWWL475ANCNFSM6AAAAAAUCWUNEY).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

MatteoRaso committed:
Added a test for positional args (PR-23061)

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -2286,6 +2286,25 @@ class vectorize:
+            pyfunc = np._NoValue
+            if p_temp is not None and p_temp is not np._NoValue:
+                o_temp = otypes
+                otypes = p_temp
```
What if someone has given really wrong input, like `vectorize(1, otypes=[...])`. I fear that for the shifting, one needs to do a `*args, **kwargs` and re-call (or re-init) with `None, *args, **kwargs` so that python can remove duplications...

MatteoRaso replied:
>What if someone has given really wrong input

That would be assuming that the user is trying to use vectorize without reading the documentation, which I don't think is right. Anyways, from tinkering with the code while trying to solve this, it seems that what would happen is that vectorize would raise ```ValueError("Invalid otype specification")```. There might be some specific incorrect input that avoids this, but I don't think that's something we have to be concerned about.

MatteoRaso committed:
@vectorize now requires arguments to specify keywords

This reverses commit 7a2ded1522305

MatteoRaso commented:
Any updates to this PR?

mattip commented:
I am still uncomfortable with the difference when using the decorator with arguments and without. I don't like the heuristics used to distinguish between the cases: I think that could fail in edge cases and the result would be hard to debug. Why do we need to reinvent the wheel here: this seems like it should be a solved problem.

MatteoRaso commented:
>I don't like the heuristics used to distinguish between the cases

Are you talking about the chain of nested if-else statements? I removed those and decided to make it so that the user has to specify the keywords for the arguments.

mattip commented:
> the user has to specify the keywords for the arguments.

But the signature accepts positional arguments as well, and making it keyword-only is a breaking change.

MatteoRaso commented:
>But the signature accepts positional arguments as well, and making it keyword-only is a breaking change.

That's a good point, but the consensus seems to be to insist on keyword arguments. Maybe we should make a post on the mailing list so everybody can vote on this, since it seems to be a controversial decision.

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -2267,18 +2269,44 @@ class vectorize:
+    def __init__(self, pyfunc=np._NoValue, otypes=None, doc=None,
+                 excluded=None, cache=False, signature=None):
+
+        if (pyfunc != np._NoValue) and (not callable(pyfunc)):
```
@mattip - I think is actually fairly safe for the decorator - if one passes in positional arguments, it is not very likely that the one for `otypes` is going to be callable.

That said, one could make it keyword only and deprecate the positional version, something like
```
def __init__(pyfunc=None, *args, otypes=None, ...):
    if args:
        warn(deprecation-warning)
        kwargs = {name : value
                  for (name, value, default) in [('otypes', otypes, None),
                                                 ...]
                  if value is not default}
        self._init_with_posargs(pyfunc, *args, **kwargs)

def _init_with_posargs(pyfunc=None, otypes=None, doc=None,
                 excluded=None, cache=False, signature=None):
    self._init(pyfunc, otypes=otypes, doc=doc, ...)
```

mattip commented:
I don't want to hold this up if others think the API is unambiguous. It does need a release note.

MatteoRaso committed:
Added release note (#23061)

MatteoRaso commented:
>I don't want to hold this up if others think the API is unambiguous. It does need a release note.

Ok, I added one. I've never written a release note before, so please look over it to make sure it's up to par.

MatteoRaso commented:
Any update to this?

rgommers commented:
> That's a good point, but the consensus seems to be to insist on keyword arguments. Maybe we should make a post on the mailing list so everybody can vote on this, since it seems to be a controversial decision.

Making the decorator version keyword-only seems perfectly fine to me. It is not backwards-incompatible, since as the release note points out the non-decorator usage keeps accepting positional-only arguments. And there's a test that positional args to the decorator version raise. So I think this point has been addressed.

mattip commented:
Thanks @MatteoRaso

MatteoRaso commented:
No problem.

------- Original Message -------
On Sunday, March 26th, 2023 at 1:44 AM, Matti Picus ***@***.***> wrote:

> Thanks ***@***.***(https://github.com/MatteoRaso)
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/numpy/numpy/pull/23061#issuecomment-1484001964), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AIDGW6TKE4HFLUI2FT6PPZDW57JSPANCNFSM6AAAAAAUCWUNEY).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

WarrenWeckesser commented:
It looks like this broke some SciPy tests.

The new version of `vectorize` doesn't work with a callable created with `functools.partial`.  For example, with numpy 1.24.2, this script

```
from functools import partial
import numpy as np


def foo(x, y):
    return x + y


boo = partial(foo, 3)
print(boo(1))

vboo = np.vectorize(boo)
print(vboo(1))
```
prints
```
4
4
```
With the main development branch, the `vectorize` call raises an exception:

```
4
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File ~/code-snippets/python/numpy/check_vectorize_with_partial.py:14
     11 boo = partial(foo, 3)
     12 print(boo(1))
---> 14 vboo = np.vectorize(boo)
     15 print(vboo(1))

File ~/py3.10.8/lib/python3.10/site-packages/numpy/lib/function_base.py:2299, in vectorize.__init__(self, pyfunc, otypes, doc, excluded, cache, signature)
   2297 self.signature = signature
   2298 if pyfunc != np._NoValue:
-> 2299     self.__name__ = pyfunc.__name__
   2301 self._ufunc = {}    # Caching to improve default performance
   2302 self._doc = None

AttributeError: 'functools.partial' object has no attribute '__name__'

```

mattip commented:
I proposed reverting the merge of this PR in #23466

MatteoRaso commented:
> It looks like this broke some SciPy tests.
> 
> The new version of `vectorize` doesn't work with a callable created with `functools.partial`. For example, with numpy 1.24.2, this script
> 
> ```
> from functools import partial
> import numpy as np
> 
> 
> def foo(x, y):
>     return x + y
> 
> 
> boo = partial(foo, 3)
> print(boo(1))
> 
> vboo = np.vectorize(boo)
> print(vboo(1))
> ```
> 
> prints
> 
> ```
> 4
> 4
> ```
> 
> With the main development branch, the `vectorize` call raises an exception:
> 
> ```
> 4
> ---------------------------------------------------------------------------
> AttributeError                            Traceback (most recent call last)
> File ~/code-snippets/python/numpy/check_vectorize_with_partial.py:14
>      11 boo = partial(foo, 3)
>      12 print(boo(1))
> ---> 14 vboo = np.vectorize(boo)
>      15 print(vboo(1))
> 
> File ~/py3.10.8/lib/python3.10/site-packages/numpy/lib/function_base.py:2299, in vectorize.__init__(self, pyfunc, otypes, doc, excluded, cache, signature)
>    2297 self.signature = signature
>    2298 if pyfunc != np._NoValue:
> -> 2299     self.__name__ = pyfunc.__name__
>    2301 self._ufunc = {}    # Caching to improve default performance
>    2302 self._doc = None
> 
> AttributeError: 'functools.partial' object has no attribute '__name__'
> ```

Hmm, it feels like an oversight for `functools.partial` to not have a `__name__` or `__doc__` attribute. Instead of adding a check to see if the input is a partial, I think it'd be better if I open a PR to add these attributes to the `partial` class. That being said, it'd delay this PR even longer, maybe by a lot. Is that okay with everybody?

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -2267,18 +2269,44 @@ class vectorize:
         self.cache = cache
         self.signature = signature
-        self._ufunc = {}    # Caching to improve default performance
+        if pyfunc != np._NoValue:
```
I would either remove this part in recreating the PR - the name didn't use to get copied, so it doesn't have to now - or add `and hasattr(pyfunc, '__name__')`

WarrenWeckesser replied:
> ...the name didn't use to get copied, so it doesn't have to now

One of the motivations for this PR was to close https://github.com/numpy/numpy/issues/23021, so I think your second suggestion--add `and hasattr(pyfunc, '__name__')`--is the better option.

mhvk commented:
I'd make things work now and possibly adjust after `functools.partial` has been adjusted - see in-line comment.
